ext/TensorKitChainRulesCoreExt/constructors.jl

```diff
@@ -4,34 +4,34 @@
 @non_differentiable TensorKit.isometry(args...)
 @non_differentiable TensorKit.unitary(args...)
 
-function ChainRulesCore.rrule(::Type{<:TensorMap}, d::DenseArray, args...)
+function ChainRulesCore.rrule(::Type{<:TensorMap}, d::DenseArray, args...; kwargs...)
     function TensorMap_pullback(Δt)
         ∂d = convert(Array, unthunk(Δt))
         return NoTangent(), ∂d, ntuple(_ -> NoTangent(), length(args))...
     end
-    return TensorMap(d, args...), TensorMap_pullback
+    return TensorMap(d, args...; kwargs...), TensorMap_pullback
 end
 
 function ChainRulesCore.rrule(::typeof(Base.copy), t::AbstractTensorMap)
     copy_pullback(Δt) = NoTangent(), Δt
     return copy(t), copy_pullback
 end
 
-# this rule does not work for generic symmetries, as we currently have no way to
-# project back onto the symmetric subspace
 function ChainRulesCore.rrule(::typeof(Base.convert), T::Type{<:Array},
-                              t::TrivialTensorMap)
+                              t::AbstractTensorMap)
     A = convert(T, t)
     function convert_pullback(ΔA)
-        ∂t = TensorMap(unthunk(ΔA), codomain(t), domain(t))
+        # use constructor to (unconditionally) project back onto symmetric subspace
+        ∂t = TensorMap(unthunk(ΔA), codomain(t), domain(t); tol=Inf)
         return NoTangent(), NoTangent(), ∂t
     end
     return A, convert_pullback
 end
 
 function ChainRulesCore.rrule(::typeof(Base.convert), ::Type{Dict}, t::AbstractTensorMap)
     out = convert(Dict, t)
-    function convert_pullback(c)
+    function convert_pullback(c′)
+        c = unthunk(c′)
         if haskey(c, :data) # :data is the only thing for which this dual makes sense
```

Comment:
How about `dual = typeof(out)(:data => c[:data])` ?

Reply:
I don't think that works, because then the spaces are missing in the dictionary-to-tensormap converter. Maybe the comment is a bit misleading -- all fields in the dictionary are required, but Zygote tends to drop fields that do not contribute (i.e. codomain and domain). Thus, this uses the dictionary from the forwards pass with the data from the backwards pass.

Reply:
Ok. I wanted to avoid copying the `data` from `out`, but probably `copy(out)` is a shallow copy that does not duplicate `out[:data]`?

Reply:
Indeed, it should really only duplicate the pointer, which I think is acceptable. On top of that, I am not sure this rule is ever used/useful anyways. I think I copied this at some point from maarten, but presumably that was also only introduced for testing purposes, as it also suffers from the weird interplay of "inner product on the parameters" for non-abelian symmetries.